### ISSUE TITLE
Batch size

### ISSUE THREAD

**Issue by stefanpeidli** [stefanpeidli/GoNet#15]
I think I missed something:

Let's say we have a data set of 1000 boards to be learned. Atm **a single** game is propagated through the network and then weights are adjusted. This is the way Heining did it (which is actually pretty terrible). This method corresponds to a **batch size of 1**. We can see [here](https://stats.stackexchange.com/questions/153531/what-is-batch-size-in-neural-network) how a batch size of 1 performs, compared to a bigger batch size: Not so good.

Thus we will introduce Batch Size as parameter for our Networks (TODO).

**Comment by stefanpeidli** [comment]
Batch size has now been introduced:
`splitintobatches(self,trainingdata,batchsize)` splits a TrainingData Set into batchs.
`Learnpropagatebatch(self, eta, batch)` propagates the batch, accumulating gradients, summing them up and **then** applies them to the weights.

Pro: Seems to accelerate learning in some way
Contra: Takes much longer than non-batch learning (Why?) EDIT: this is due to a bug in Learningpropagate method, which has been fixed.

**Comment by stefanpeidli** [comment]
First results show that introducing batchs might not necessarily be a very good thing to do.

A test with Dan_10 games, eta=0.001, epochs=2, batchsize=700 leads to the following results:

![image](https://user-images.githubusercontent.com/24811160/33632209-424b707c-da0d-11e7-9773-277f820a306d.png)

The first line is the initial error (untrained Network) and the time it took to propagate the whole set.
The second line is the error after batch-learning, the time it took and the time for spliting the set into batchs.
(Between the tests, here, the weights are resetted to the same state, allowing better comparison.)
The third one is Learning with a batch size of 1, the way we started implementing the Learning algorithm.

We see that **the Batch-method is marginally faster, but performs quite a bit worse** when it comes to the error (not supprising, batch size 1 has highest accuracy in most cases, but is slower usually). Also we observe that spliting a Dataset into batches is done really fast.

**Comment by paddyooo** [comment]
I would feel like a smaller batchsize and a higher epoch size would feel better for that amout of data

**Comment by stefanpeidli** [comment]
The right choice of the batch size is still something mystical. I think we will have to find an "optimal" choice by testing.

Smaller test size sounds good, higher epoch too. But this test above was intentionally done with a big batch size (in order to compare big and small batch size 700 : 1, and the epochs small, because I did not want to sit there waiting forever 😄

**Comment by Beno71** [comment]
New "problem" here.
Our training will be very inefficient for batch_size one.
In general bigger batch_sizes will boost the training speed a lot I think. Two reasons:
1 Less backpropagation
2 Less dictionaries in batch extraction with more data inside
